File: 152/Solution.java

public class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        
        int result = nums[0];
        int maxProduct = nums[0];
        int minProduct = nums[0];
        
        for (int i = 1; i < nums.length; i++) {
            // 当遇到负数时，最大值和最小值会互换
            if (nums[i] < 0) {
                int temp = maxProduct;
                maxProduct = minProduct;
                minProduct = temp;
            }
            
            // 更新最大值和最小值
            maxProduct = Math.max(nums[i], nums[i] * maxProduct);
            minProduct = Math.min(nums[i], nums[i] * minProduct);
            
            // 更新全局最大值
            result = Math.max(result, maxProduct);
        }
        
        return result;
    }
}

File: 152/Solution.cpp
#include <vector>
#include <algorithm>


class Solution {
public:
    int maxProduct(std::vector<int>& nums){
        if (nums.empty()) return 0;
        int res = nums[0];
        int max = nums[0]; 
        int min = nums[0];
        for (int i = 1 ; i < nums.size() ; i++){
            if (nums[i] < 0){
                swap(max,min);
            }
            max = std::max(nums[i],nums[i] * max);
            min = std::min(nums[i],nums[i] * min);

            res = std::max(res,max);
        }
        return res;
    }
};